File: sol/parallel_search-duplicate.f90
PROGRAM search  
  implicit none
  include "mpif.h"
  integer, parameter ::  N=300
  integer i, target ! local variables
  integer b(N)      ! the entire array of integers
  integer myrank,size,ierr
  character*15 outfilename
  character*4 rankchar

  call MPI_Init(ierr)
  call MPI_Comm_rank(MPI_COMM_WORLD,myrank,ierr)
  call MPI_Comm_size(MPI_COMM_WORLD,size,ierr)

  ! generate the name of the output file
  ! all mpi tasks must write to a different file
  write(rankchar,'(i4.4)') myrank
  outfilename="found.data_" // rankchar
 
  ! File b.data has the target value on the first line
  ! The remaining 300 lines of b.data have the values for the b array
  open(unit=10,file="b.data")     

  ! File found.data will contain the indices of b where the target is
  open(unit=11,file=outfilename)

  ! Read in the target
  read(10,*) target

  ! Read in b array 

  do i=1,N
     read(10,*) b(i)
  end do

  ! Search the b array and output the target locations

  do i=1,N
     if (b(i) == target) then
        write(11,*) i
     end if
  end do

  write(*,*) "Processor ",myrank," of ",size,": Finished!"

  call MPI_Finalize(ierr)

END PROGRAM search

File: lab1/sol/parallel_search-duplicate.c
#include <stdlib.h>
#include <stdio.h>
#include <mpi.h>

int main (int argc, char *argv[]) {
  const int N=300;
  int i,target;
  int b[N];
  FILE *infile,*outfile;
  int myrank, size;

  char outfilename[16] ;


  MPI_Init(&argc, &argv);                 /* Initialize MPI       */
  MPI_Comm_rank(MPI_COMM_WORLD, &myrank); /* Get my rank          */
  MPI_Comm_size(MPI_COMM_WORLD, &size);   /* Get the total
					     number of processors */

  /* generate the name of the output file
     all mpi tasks must write to a different file */
  sprintf(outfilename,"found.data_%d",myrank);

  /* File b.data has the target value on the first line
     The remaining 300 lines of b.data have the values for the b array */

  infile = fopen("b.data","r" ) ;
  outfile = fopen(outfilename,"w") ;
    
  /* read in target */
  fscanf(infile,"%d", &target);

  /* read in b array */
  for(i=0;i<N;i++) {
    fscanf(infile,"%d", &b[i]);
  }
  fclose(infile);

  /* Search the b array and output the target locations */

  for(i=0;i<N;i++) {
    if( b[i] == target) {
      fprintf(outfile,"%d\n",i+1);
    }
  }
  fclose(outfile);

  printf("Processor %d of %d: finished!\n", myrank, size);
  MPI_Finalize();                         /* Terminate MPI        */

  return 0;
}
